lib/extractor/oci-archive/layer.ts

```diff
@@ -0,0 +1,145 @@
+import { createReadStream } from "fs";
+import * as gunzip from "gunzip-maybe";
+import { PassThrough } from "stream";
+import { extract, Extract } from "tar-stream";
+import { streamToJson } from "../../stream-utils";
+import { extractImageLayer } from "../layer";
+import {
+  ExtractAction,
+  ExtractedLayers,
+  OciArchiveManifest,
+  OciImageIndex,
+  OciManifestInfo,
+} from "../types";
+
+/**
+ * Retrieve the products of files content from the specified oci-archive.
+ * @param ociArchiveFilesystemPath Path to image file saved in oci-archive format.
+ * @param extractActions Array of pattern-callbacks pairs.
+ * @returns Array of extracted files products sorted by the reverse order of the layers from last to first.
+ */
+export async function extractArchive(
+  ociArchiveFilesystemPath: string,
+  extractActions: ExtractAction[],
+): Promise<{
+  layers: ExtractedLayers[];
+  manifest: OciArchiveManifest;
+}> {
+  return new Promise((resolve, reject) => {
+    const tarExtractor: Extract = extract();
+
+    const layers: Record<string, ExtractedLayers> = {};
+    const manifests: Record<string, OciArchiveManifest> = {};
+    let imageIndex: OciImageIndex | undefined;
+
+    tarExtractor.on("entry", async (header, stream, next) => {
+      if (header.type === "file") {
+        if (isImageIndexFile(header.name)) {
+          imageIndex = await streamToJson<OciImageIndex>(stream);
+        } else {
+          const jsonStream = new PassThrough();
+          const layerStream = new PassThrough();
+          stream.pipe(jsonStream);
+          stream.pipe(layerStream);
+
+          const promises = [
+            streamToJson(jsonStream).catch(() => undefined),
+            extractImageLayer(layerStream, extractActions).catch(
+              () => undefined,
+            ),
+          ];
+          const [manifest, layer] = await Promise.all(promises);
+
+          // header format is /blobs/hash_name/hash_value
+          // we're extracting hash_name:hash_value format to match manifest digest
+          const headerParts = header.name.split("/");
```

Comment:
[Minor point]
To improve the code readability, maybe we could push complexity about how to get `hashName` and `hashValue` in a function. What do you think?